backend/submodule/clock.py

```diff
@@ -0,0 +1,139 @@
+
+    def get_clocking_resources(self, total_clock_available, total_pll_available):
+        total_clock_used = sum(1 for clock in self.clocks if clock.enable and clock.source == Source.IO)
+        total_pll_used = sum(1 for clock in self.clocks if clock.enable and clock.source != Source.IO)
+
+        used_clocks = total_clock_available - total_clock_used
+        used_plls = total_pll_available - total_pll_used
+        used_clocks_percentage = used_clocks / total_clock_available * 100
+        used_plls_percentage = used_plls / total_pll_available * 100
+        used_clocks_power = 0.0  # todo, figure out the formula to calculate the power
+        used_plls_power = 0.0  # todo, figure out the formula to calculate the power
+        return used_clocks, used_plls, used_clocks_percentage, used_plls_percentage, used_clocks_power
+
+    def get_clocks(self):
+        return self.clocks
+    
+    def get_clock(self, idx):
+        # Check if the index is within the valid range
+        if 0 <= idx < len(self.clocks):
+            # Return the clock at the specified index
+            return self.clocks[idx]
+        else:
+            raise ValueError("Invalid index. Clock doesn't exist at the specified index.")
+
+    def add_clock(self, clock):
+        # Check if the clock already exists based on the ID
+        if any(existing_clock.description == clock.description for existing_clock in self.clocks):
+            raise ValueError("Clock description already exists in the list of clocks.")
+        # Add the clock to the list
+        self.clocks.append(clock)
+
+    def delete_clock(self, idx):
+        # Check if the index is within the valid range
+        if 0 <= idx < len(self.clocks):
+            # Remove the clock at the specified index
+            deleted_clock = self.clocks[idx]
+            del self.clocks[idx]
+            return deleted_clock
+        else:
+            raise ValueError("Invalid index. Clock doesn't exist at the specified index.")
+
+    def update_clock(self, idx, clock_data):
+        # Check if the provided index is valid
+        if 0 <= idx < len(self.clocks):
+            self.clocks[idx].enable = clock_data['enable']
+            self.clocks[idx].description = clock_data['description']
+            self.clocks[idx].source = clock_data['source']
+            self.clocks[idx].port = clock_data['port']
+            self.clocks[idx].frequency = clock_data['frequency']
+            self.clocks[idx].state = clock_data['state']
+        else:
+            raise ValueError("Invalid index provided for updating the clock.")
+    
+    def compute_clocks_output_power(self):
+        # Compute the total power consumption of all clocks
+        total_power = 0
+        for clock in self.clocks:
+            total_power += clock.compute_dynamic_power()
+        return total_power
+
+if __name__ == '__main__':
+    # manual test code
+    clock = Clock(enable=True)
+    clock2 = Clock(enable=True)
+    print(clock)
```

Comment:
Please replace with the logger class in future PR. We don't want print statements on the console.

Reply:
Noted.